src/server.cpp

```diff
@@ -1603,49 +1606,77 @@ void Server::SendShowFormspecMessage(session_t peer_id, const std::string &forms
 	Send(&pkt);
 }
 
-// Spawns a particle on peer with peer_id
-void Server::SendSpawnParticle(session_t peer_id, u16 protocol_version,
-	const ParticleParameters &p)
+void Server::SendSpawnParticles(session_t peer_id, const std::vector<ParticleParameters> &particles)
 {
 	static thread_local const float radius =
-			g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
-
-	if (peer_id == PEER_ID_INEXISTENT) {
-		std::vector<session_t> clients = m_clients.getClientIDs();
-		const v3f pos = p.pos * BS;
-		const float radius_sq = radius * radius;
-
-		for (const session_t client_id : clients) {
-			RemotePlayer *player = m_env->getPlayer(client_id);
-			if (!player)
-				continue;
+		g_settings->getS16("max_block_send_distance") * MAP_BLOCKSIZE * BS;
+	const float radius_sq = radius * radius;
 
-			PlayerSAO *sao = player->getPlayerSAO();
-			if (!sao)
-				continue;
+	RemoteClient *client = m_clients.getClientNoEx(peer_id, CS_Active);
+	if (!client)
+		return;
 
-			// Do not send to distant clients
-			if (sao->getBasePosition().getDistanceFromSQ(pos) > radius_sq)
-				continue;
+	PlayerSAO *sao = getPlayerSAO(peer_id);
+	if (!sao)
+		return;
 
-			SendSpawnParticle(client_id, player->protocol_version, p);
+	RemotePlayer *player = m_env->getPlayer(peer_id);
+	assert(player);
+
+	std::ostringstream oss(std::ios::binary);
+	for (const auto &particle : particles) {
+		if (sao->getBasePosition().getDistanceFromSQ(particle.pos * BS) > radius_sq)
+			continue; // out of range
+
+		particle.serialize(oss, player->protocol_version);
+		if (player->protocol_version < 50) {
+			// Client only supports TOCLIENT_SPAWN_PARTICLE,
+			// so turn the written particle into a packet immediately
+			NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE, oss.tellp(), peer_id);
+			pkt.putRawString(oss.str());
+			Send(&pkt);
+			oss.clear();
 		}
-		return;
 	}
-	assert(protocol_version != 0);
 
-	NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE, 0, peer_id);
+	if (oss.tellp() == 0)
+		return; // no batch to send
 
+	// Client supports TOCLIENT_SPAWN_PARTICLE_BATCH
+	assert(player->protocol_version >= 50);
+	std::string compressed_data;
```

Comment:
Why do you need this copy? `compressed_data_oss` could be used directly to feed `NetworkPacket` with the necessary data.

Reply:
What would you suggest? As far as I know there is no good C++ 17 way to get a view into the underlying buffer of a stringstream without copying. C++ 20 introduces the `.view()` method for that, but in C++ 17 we only have `.str()`. There is `.rdbuf()` but it is not obvious to me from the specification that this would be guaranteed to give us the entire buffer.

I also don't think this copy is very significant in the bigger picture.

Reply:
```C++
	std::ostringstream compressed_data;
	compressZstd(oss.str(), compressed_data);

	NetworkPacket pkt(TOCLIENT_SPAWN_PARTICLE_BATCH,
			4 + compressed_data.tellp(), player->getPeerId());
	pkt.putLongString(compressed_data.str());
	Send(&pkt);
```

Reply:
`compressZstd` only takes strings at the moment, so I've left it at that. I think this is still doing same number of copies (first `oss` / `particle_data` is copied to get it out of the string buffer and into zstd compression, then the compressed data is copied afterwards; the assignment `compressed_data = compressed_data_oss.str()` should be a move assignment), but it is a bit nicer to read.

Reply:
> `compressZstd` only takes strings at the moment

it takes a `std::string_view`